cpp/src/arrow/dataset/partition.h

```diff
@@ -18,87 +18,114 @@
-//                            std::shared_ptr<Scalar>* output) const = 0;
-// };
-
-// class PartitionField {
-//  public:
-
-//  private:
-//   std::string field_name_;
-// };
-
-// ----------------------------------------------------------------------
-// Partition identifiers
-
-/// \brief A partition level identifier which can be used
-///
-/// TODO(wesm): Is this general enough? What other kinds of partition
-/// keys exist and do we need to support them?
-class PartitionKey {
- public:
-  const std::vector<std::string>& fields() const { return fields_; }
-  const std::vector<std::shared_ptr<Scalar>>& values() const { return values_; }
+// Partition schemes
 
- private:
-  std::vector<std::string> fields_;
-  std::vector<std::shared_ptr<Scalar>> values_;
+struct ARROW_DS_EXPORT UnconvertedKey {
+  std::string name, value;
 };
 
-/// \brief Intermediate data structure for data parsed from a string
-/// partition identifier.
+/// \brief Helper function for the common case of combining partition information
+/// consisting of equality expressions into a single conjunction expression
+ARROW_DS_EXPORT
+Status ConvertPartitionKeys(const std::vector<UnconvertedKey>& keys, const Schema& schema,
+                            std::shared_ptr<Expression>* out);
+
+/// \brief Interface for parsing partition expressions from string partition
+/// identifiers.
 ///
-/// For example, the identifier "foo=5" might be parsed with a single
-/// "foo" field and the value 5. A more complex identifier might be
-/// written as "foo=5,bar=2", which would yield two fields and two
-/// values.
+/// For example, the identifier "foo=5" might be parsed to an equality expression
+/// between the "foo" field and the value 5.
 ///
 /// Some partition schemes may store the field names in a metadata
 /// store instead of in file paths, for example
 /// dataset_root/2009/11/... could be used when the partition fields
 /// are "year" and "month"
-struct PartitionKeyData {
-  std::vector<std::string> fields;
-  std::vector<std::shared_ptr<Scalar>> values;
-};
-
-// ----------------------------------------------------------------------
-// Partition schemes
-
-/// \brief
+///
+/// Paths are consumed from left to right. Paths must be relative to
+/// the root of a partition; path prefixes must be removed before passing
+/// the path to a scheme for parsing.
 class ARROW_DS_EXPORT PartitionScheme {
  public:
   virtual ~PartitionScheme() = default;
 
   /// \brief The name identifying the kind of partition scheme
   virtual std::string name() const = 0;
 
-  virtual bool PathMatchesScheme(const std::string& path) const = 0;
+  /// \brief Parse a path into a partition expression
+  ///
+  /// \param[in] path the partition identifier to parse
+  /// \param[out] unconsumed a suffix of path which was not consumed
+  /// \param[out] out the parsed expression
+  virtual Status Parse(const std::string& path, std::string* unconsumed,
```

Comment:
I recommend that we drop the `std::string* unconsumed` parameter from the `PartitionScheme::Parse` signature. The current draft requires that PartitionScheme stops and indicate when they can't consume a part of the path (string). I don't think this is needed, i.e. we should stick to `Parse(string) -> expression`.

My observations are the following:

1. Users should be exposed the simplest interface.
2. Users will use 0 or 1 partition scheme, rarely will they have more than 1 (for a single data source). This make chaining a low priority item.
3. It's debatable if `HivePartitionScheme.Parse("/a=1/b=2/blah/c=3")` returns `{a: 1, b:2, c:3}` or `{a: 1: b:2}`. Impala/Hive doc is not clear on this, I suspect it's impossible. We can expose a flag in the constructor that dictate the desired behaviour.
4. The natural expectation is that chaining would just concatenate their extracted partitions from the full path each time, not consume the leftover of the previous scheme. If the user wants a complex scheme, he can implement one.
```
partitions = map(schemes, lambda s: s.Parse(path))
return fold(partitions, expr(true), lambda e_1, e_2: e_1 && e_2)
```

We're making the interface more complex (partial parsing chaining) for very little benefit. Instead, we should offer very basic PartitionScheme, e.g. RegularExpression and Lambda.

Reply:
Although `(path)-> expression, unconsumed` is a slightly more complicated interface than `(path)-> expression` I think it is more important make path scanning unsurprising. Strict left to right consumption ensures that two chained schemes are not parsing information which wasn't intended for them. If we don't provide guarantees about how the path is consumed users have to address this by tinkering with their regular expressions (not catastrophic, but it seems more fragile IMHO)

We could remove chaining altogether but from a design perspective it seems more favorable to make modular components which can be straightforwardly composed. I think we're going to want the ability to stack schemes

Reply:
Refactored to `(path)->expression`, removed chain partition scheme